pkg/scheduler/plugins.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scheduler
+
+import schedulerapi "k8s.io/kubernetes/pkg/scheduler/apis/config"
+
+// mergePlugins merge default plugins into custom plugins.
+func mergePlugins(customPlugins, defaultPlugins *schedulerapi.Plugins) *schedulerapi.Plugins {
+	if customPlugins == nil {
+		return defaultPlugins
+	}
+
+	if defaultPlugins == nil {
+		return customPlugins
+	}
+
+	newPlugins := &schedulerapi.Plugins{}
+	newPlugins.QueueSort = mergeDefaultPluginSet(customPlugins.QueueSort, defaultPlugins.QueueSort)
+	newPlugins.PreFilter = mergeDefaultPluginSet(customPlugins.PreFilter, defaultPlugins.PreFilter)
+	newPlugins.Filter = mergeDefaultPluginSet(customPlugins.Filter, defaultPlugins.Filter)
+	newPlugins.PostFilter = mergeDefaultPluginSet(customPlugins.PostFilter, defaultPlugins.PostFilter)
+	newPlugins.Score = mergeDefaultPluginSet(customPlugins.Score, defaultPlugins.Score)
+	newPlugins.Reserve = mergeDefaultPluginSet(customPlugins.Reserve, defaultPlugins.Reserve)
+	newPlugins.Permit = mergeDefaultPluginSet(customPlugins.Permit, defaultPlugins.Permit)
+	newPlugins.PreBind = mergeDefaultPluginSet(customPlugins.PreBind, defaultPlugins.PreBind)
+	newPlugins.Bind = mergeDefaultPluginSet(customPlugins.Bind, defaultPlugins.Bind)
+	newPlugins.PostBind = mergeDefaultPluginSet(customPlugins.PostBind, defaultPlugins.PostBind)
+	newPlugins.Unreserve = mergeDefaultPluginSet(customPlugins.Unreserve, defaultPlugins.Unreserve)
+
+	return newPlugins
+}
+
+func mergeDefaultPluginSet(customPluginSet, defaultPluginSet *schedulerapi.PluginSet) *schedulerapi.PluginSet {
+	if customPluginSet == nil {
+		customPluginSet = &schedulerapi.PluginSet{}
+	}
+
+	if defaultPluginSet == nil {
+		defaultPluginSet = &schedulerapi.PluginSet{}
+	}
+
+	disabledPlugins := map[string]schedulerapi.Plugin{}
+	for i := range customPluginSet.Disabled {
+		disabledPlugin := customPluginSet.Disabled[i]
+		disabledPlugins[disabledPlugin.Name] = disabledPlugin
+	}
+
+	enabledPlugins := []schedulerapi.Plugin{}
+
+	_, ok := disabledPlugins["*"]
+	if !ok {
+		for i := range defaultPluginSet.Enabled {
```

Comment:
why not `for _, p := range`?

Reply:
FYI: https://github.com/golang/go/wiki/CommonMistakes#using-reference-to-loop-iterator-variable

It would be better to use this form when appending the elements of a slice to another.

Reply:
This is true for pointers, but these are values. You are producing copies in the loop anyways.

Reply:
> This is true for pointers, but these are values. You are producing copies in the loop anyways.

Yes, you are right.